### PR TITLE
usbdev: add USB cdcncm device driver

### PR DESCRIPTION
**Pull request by zhhyu7**
## Summary
Add USB cdcncm device driver.

The compilation and verification commands are shown below:
./tools/configure.sh sim:usbdev
make -j
sudo ./nuttx
nsh> conn 2
nsh> dhcpd_start eth1

## Testing
sim:usbdev with https://github.com/xairy/raw-gadget.git on ubuntu.

### CODE REVIEW

File: boards/sim/sim/sim/src/sim_composite.c [hunk 1]
```diff
@@ -210,6 +211,60 @@ static void *board_composite1_connect(int port)
...
+ *   A non-NULL handle value is returned on success.  NULL is returned on
+ *   any failure.
+ *
+ ****************************************************************************/
+
+static void *board_composite2_connect(int port)
+{
+  struct composite_devdesc_s dev[1];
+  int ifnobase = 0;
+  int strbase = COMPOSITE_NSTRIDS - 1;
```
Comment (anchao) [inline boards/sim/sim/sim/src/sim_composite.c:234]:
```suggestion

```
move after line 237

File: drivers/usbdev/cdcncm.c [hunk 2]
```diff
@@ -0,0 +1,2541 @@
...
+      EP_SUBMIT(self->epbulkout, self->rdreq);
+      leave_critical_section(flags);
+    }
+
+  /* Check if a packet transmission just completed.  If so, call
+   * cdcncm_txdone. This may disable further Tx interrupts if there
+   * are no pending transmissions.
+   */
+
+  if (self->txdone)
```
Comment (anchao) [inline drivers/usbdev/cdcncm.c:908]: self->txdone
```suggestion
  flags = enter_critical_section();
  if (self->txdone)
```

File: drivers/usbdev/cdcncm.c [hunk 3]
```diff
@@ -0,0 +1,2541 @@
...
+ *
+ ****************************************************************************/
+
+static FAR netpkt_t *cdcncm_recv(FAR struct netdev_lowerhalf_s *dev)
+{
+  FAR struct cdcncm_driver_s *self;
+  FAR netpkt_t *pkt;
+
+  self = container_of(dev, struct cdcncm_driver_s, dev);
+  pkt  = netpkt_remove_queue(&self->rx_queue);
```
Comment (anchao) [inline drivers/usbdev/cdcncm.c:1056]:
```suggestion
  return netpkt_remove_queue(&self->rx_queue);
```

File: drivers/usbdev/cdcncm.c [hunk 4]
```diff
@@ -0,0 +1,2541 @@
...
+#if defined(CONFIG_NET_MCASTGROUP) || defined(CONFIG_NET_ICMPv6)
+static int cdcncm_addmac(FAR struct netdev_lowerhalf_s *dev,
+                         FAR const uint8_t *mac)
+{
+  FAR struct cdcncm_driver_s *priv =
+    container_of(dev, struct cdcncm_driver_s, dev);
+
+  /* Add the MAC address to the hardware multicast routing table */
+
+  UNUSED(priv); /* Not yet implemented */
```
Comment (anchao) [inline drivers/usbdev/cdcncm.c:1085]:
```suggestion

```

File: drivers/usbdev/cdcncm.c [hunk 5]
```diff
@@ -0,0 +1,2541 @@
...
+#ifdef CONFIG_NET_MCASTGROUP
+static int cdcncm_rmmac(FAR struct netdev_lowerhalf_s *dev,
+                        FAR const uint8_t *mac)
+{
+  FAR struct cdcncm_driver_s *priv =
+    container_of(dev, struct cdcncm_driver_s, dev);
+
+  /* Add the MAC address to the hardware multicast routing table */
+
+  UNUSED(priv); /* Not yet implemented */
```
Comment (anchao) [inline drivers/usbdev/cdcncm.c:1115]:
```suggestion

```

File: drivers/usbdev/cdcncm.c [hunk 6]
```diff
@@ -0,0 +1,2541 @@
...
+  switch (cmd)
+    {
+      /* Add cases here to support the IOCTL commands */
+
+      default:
+        nerr("ERROR: Unrecognized IOCTL command: %d\n", cmd);
+        return -ENOTTY;  /* Special return value for this case */
+    }
+
+  return OK;
```
Comment (anchao) [inline drivers/usbdev/cdcncm.c:1154]:
```suggestion
  return -ENOTTY;  /* Special return value for this case */
```

File: drivers/usbdev/cdcncm.c [hunk 7]
```diff
@@ -0,0 +1,2541 @@
...
+ *   Set the device configuration by allocating and configuring endpoints and
+ *   by allocating and queue read and write requests.
+ *
+ ****************************************************************************/
+
+static int cdcncm_setconfig(FAR struct cdcncm_driver_s *self, uint8_t config)
+{
+  struct usb_epdesc_s epdesc;
+  bool is_high_speed = self->usbdev.speed == USB_SPEED_HIGH;
+  int ret = OK;
```
Comment (anchao) [inline drivers/usbdev/cdcncm.c:1301]:
```suggestion
  int ret;
```

File: drivers/usbdev/cdcncm.c [hunk 8]
```diff
@@ -0,0 +1,2541 @@
...
+        req->value[1] = MSBYTE(IFF_IS_RUNNING(self->dev.netdev.d_flags));
+        req->len[0]   = 0;
+        req->len[1]   = 0;
+        ret           = sizeof(*req);
+
+        self->notify  = NCM_NOTIFY_NONE;
+        break;
+
+      case NCM_NOTIFY_SPEED:
+        {
```
Comment (anchao) [inline drivers/usbdev/cdcncm.c:1414]:
```suggestion
```

File: drivers/usbdev/cdcncm.c [hunk 9]
```diff
@@ -0,0 +1,2541 @@
...
+    {
+      ret = ncm_notify(self);
+      if (ret > 0)
+        {
+          FAR struct usbdev_req_s *notifyreq = self->notifyreq;
+
+          notifyreq->len   = ret;
+          notifyreq->flags = USBDEV_REQFLAGS_NULLPKT;
+
+          ret = EP_SUBMIT(self->epint, notifyreq);
```
Comment (anchao) [inline drivers/usbdev/cdcncm.c:1463]:
```suggestion
          EP_SUBMIT(self->epint, notifyreq);
```

**Review by anchao — CHANGES_REQUESTED** [review]

**Review by acassis — COMMENTED** [review]
Please include that instruction from message commit log to https://nuttx.apache.org/docs/latest/platforms/sim/sim/boards/sim/index.html

File: drivers/usbdev/cdcncm.c (hunk 8 above)
Comment (zhhyu7) [reply drivers/usbdev/cdcncm.c:1414]: `FAR uint32_t *data;`
1415 have variable definitions, do I need to delete it?

**Comment by zhhyu7** [comment]
> Please include that instruction from message commit log to https://nuttx.apache.org/docs/latest/platforms/sim/sim/boards/sim/index.html

Done.

File: drivers/usbdev/cdcncm.c (hunk 9 above)
Comment (zhhyu7) [reply drivers/usbdev/cdcncm.c]: Done.

File: drivers/usbdev/cdcncm.c (hunk 7 above)
Comment (zhhyu7) [reply drivers/usbdev/cdcncm.c]: Done.

File: drivers/usbdev/cdcncm.c (hunk 6 above)
Comment (zhhyu7) [reply drivers/usbdev/cdcncm.c]: Done.

File: drivers/usbdev/cdcncm.c (hunk 5 above)
Comment (zhhyu7) [reply drivers/usbdev/cdcncm.c]: Done.

File: drivers/usbdev/cdcncm.c (hunk 4 above)
Comment (zhhyu7) [reply drivers/usbdev/cdcncm.c]: Done.

File: drivers/usbdev/cdcncm.c (hunk 3 above)
Comment (zhhyu7) [reply drivers/usbdev/cdcncm.c:1060]: Done.

File: drivers/usbdev/cdcncm.c (hunk 2 above)
Comment (zhhyu7) [reply drivers/usbdev/cdcncm.c:909]: Done.

File: boards/sim/sim/sim/src/sim_composite.c (hunk 1 above)
Comment (zhhyu7) [reply boards/sim/sim/sim/src/sim_composite.c]: Done.

**Comment by TimJTi** [comment]
Excited - can't wait for this to be ready for me to try out :-)

**Comment by zhhyu7** [comment]
@TimJTi  The current version can only be tested in sim, the physical board will have problems with the notify message due to timing problems. I will submit the driver of cdcmbim and the corresponding fix later.

File: drivers/usbdev/cdcncm.c (hunk 8 above)
Comment (anchao) [reply drivers/usbdev/cdcncm.c:1398]: ok, let us keep the bracket

File: drivers/usbdev/cdcncm.c [hunk 10]
```diff
@@ -0,0 +1,2525 @@
...
+
+#ifdef CONFIG_NET_CDCNCM
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+/* Work queue support is required. */
+
+#if !defined(CONFIG_SCHED_WORKQUEUE)
```
Comment (anchao) [inline drivers/usbdev/cdcncm.c:62]:
```suggestion
#if !defined(CONFIG_SCHED_WORKQUEUE)
```

why not set NET_CDCNCM depends on SCHED_WORKQUEUE?

**Review by anchao — APPROVED** [review]

File: drivers/usbdev/cdcncm.c (hunk 10 above)
Comment (zhhyu7) [reply drivers/usbdev/cdcncm.c]: Done.

File: boards/sim/sim/sim/src/sim_composite.c [hunk 11]
```diff
@@ -210,6 +211,61 @@ static void *board_composite1_connect(int port)
...
+  /* Endpoints */
+
+  dev[dev_idx].devinfo.epno[CDCNCM_EP_INTIN_IDX] = 5;
+  dev[dev_idx].devinfo.epno[CDCNCM_EP_BULKIN_IDX] = 6;
+  dev[dev_idx].devinfo.epno[CDCNCM_EP_BULKOUT_IDX] = 7;
+
+  /* Count up the base numbers */
+
+  ifnobase += dev[dev_idx].devinfo.ninterfaces;
+  strbase += dev[dev_idx].devinfo.nstrings;
```
Comment (anchao) [inline boards/sim/sim/sim/src/sim_composite.c:261]:
```suggestion
  ifnobase += dev[dev_idx].devinfo.ninterfaces;
  strbase += dev[dev_idx].devinfo.nstrings;
```
could be removed?

**Review by anchao — APPROVED** [review]

File: boards/sim/sim/sim/src/sim_composite.c (hunk 11 above)
Comment (zhhyu7) [reply boards/sim/sim/sim/src/sim_composite.c]: Done.

**Review by anchao — APPROVED** [review]

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by TimJTi** [comment]
> @TimJTi The current version can only be tested in sim, the physical board will have problems with the notify message due to timing problems. I will submit the driver of cdcmbim and the corresponding fix later.

@zhhyu7 - is there any reason I can't now use cdc/ncm with real hardware? I don't think I need the mbim side of things, just looking to use cdc/ncm to connect my "gadget" to Windows as RNDIS is a pain to get working on Windows!!

Or is it still sim only for some reason?

I have tried it on my board but it causes a DEBUGASSERT and don't know if it's do to a related setting I've not done or am actual issue/problem with the driver

**Comment by zhhyu7** [comment]
> > @TimJTi The current version can only be tested in sim, the physical board will have problems with the notify message due to timing problems. I will submit the driver of cdcmbim and the corresponding fix later.
> 
> @zhhyu7 - is there any reason I can't now use cdc/ncm with real hardware? I don't think I need the mbim side of things, just looking to use cdc/ncm to connect my "gadget" to Windows as RNDIS is a pain to get working on Windows!!
> 
> Or is it still sim only for some reason?
> 
> I have tried it on my board but it causes a DEBUGASSERT and don't know if it's do to a related setting I've not done or am actual issue/problem with the driver

@TimJTi cdcncm can run on hardware devices now, can you provide the backtrace related to the DEBUGASSERT? Thanks.

**Comment by zhhyu7** [comment]
> > @TimJTi The current version can only be tested in sim, the physical board will have problems with the notify message due to timing problems. I will submit the driver of cdcmbim and the corresponding fix later.
> 
> @zhhyu7 - is there any reason I can't now use cdc/ncm with real hardware? I don't think I need the mbim side of things, just looking to use cdc/ncm to connect my "gadget" to Windows as RNDIS is a pain to get working on Windows!!
> 
> Or is it still sim only for some reason?
> 
> I have tried it on my board but it causes a DEBUGASSERT and don't know if it's do to a related setting I've not done or am actual issue/problem with the driver

@TimJTi cdcncm can run on hardware devices now, can you provide the backtrace related to the DEBUGASSERT? Thanks.

**Comment by TimJTi** [comment]
@zhhyu7 This is on a board with a MicroChip SAMA5D27.

There's a DEBUGASSERT line 2380) that trips on the second call to this after USB attach, because "the request from the head of the endpoint request queue" is NULL.

Endpoints and USB are not my thing at all - but happy to try things and do further tests/debug if it helps! 

It could be IN/OUT/INT EPx assignments - I'm playing around with that at the moment.

Callstack:
```

sam_dma_interrupt@0x200260cc (/home/tim/designs/nuttx/nuttx/arch/arm/src/sama5/sam_udphs.c:2380)
sam_udphs_interrupt@0x200260cc (/home/tim/designs/nuttx/nuttx/arch/arm/src/sama5/sam_udphs.c:2976)
arm_doirq@0x200090fc (/home/tim/designs/nuttx/nuttx/arch/arm/src/armv7-a/arm_doirq.c:86)
sam_decodeirq@0x2000861c (/home/tim/designs/nuttx/nuttx/arch/arm/src/sama5/sam_irq.c:605)
arm_decodeirq@0x2000861c (/home/tim/designs/nuttx/nuttx/arch/arm/src/sama5/sam_irq.c:617)
arm_vectorirq@0x20008080 (/home/tim/designs/nuttx/nuttx/arch/arm/src/armv7-a/arm_vectors.S:235)
```

**Comment by TimJTi** [comment]
FYI - if I comment out the DEBUGASSERT in the sam usb device driver it does connect and work - I can connect to my board from a PC :-)

So this must be a very odd SAMA5-specific issue so probably not for you to worry about, although happy to still receive suggestions!!!

**Pull request merged**